File: quick_sort.py
def partition(array, low, high):
    pivot = high
    left = low
    for right in range(low, high):
        if array[right] < array[pivot]:
            array[left], array[right] = array[right], array[left]
            left += 1
    array[left], array[high] = array[high], array[left]
    return left


def quickSort(array, low, high):
    if low < high:
        pivot = partition(array, low, high)
        quickSort(array, low, pivot-1)
        quickSort(array, pivot+1, high)
 
 
data = [1, 7, 4, 1, 10, 9, -2]
print("Unsorted Array")
print(data)
 
size = len(data)
 
quickSort(data, 0, size - 1)
 
print('Sorted Array in Ascending Order:')
print(data)

File: quick_sort.c
#include <stdio.h>

void swap(int *p1, int *p2)
{
  int temp;
  temp = *p1;
  *p1 = *p2;
  *p2 = temp;
}

int partition(int arr[], int low, int high)
{
  // choose the pivot
  int pivot = arr[high];

  int left = low;

  for (int right = low+1; right < high; right++)
  {
    // If current element is smaller than the pivot
    if (arr[right] < pivot)
    {
      swap(&arr[left], &arr[right]);
      left++;
    }
  }
  swap(&arr[left], &arr[high]);
  return left;
}

// The Quicksort function Implement

void quickSort(int arr[], int low, int high)
{
  // when low is less than high
  if (low < high)
  {
    // pi is the partition return index of pivot

    int pi = partition(arr, low, high);

    // Recursion Call
    // smaller element than pivot goes left and
    // higher element goes right
    quickSort(arr, low, pi - 1);
    quickSort(arr, pi + 1, high);
  }
}

int main()
{
  int arr[] = {11, 7, 8, 9, 1, 5};
  int n = sizeof(arr) / sizeof(arr[0]);

  // Function call
  quickSort(arr, 0, n - 1);

  // Print the sorted array
  printf("Sorted Array\n");
  for (int i = 0; i < n; i++)
  {
    printf("%d ", arr[i]);
  }
  return 0;
}
